### DISCUSSION TITLE
Possibility to output the parsed OpenAPI spec?

### DISCUSSION THREAD

**Discussion by fsvenson** [Himenon/openapi-typescript-code-generator#52] [Q&A] [answered]
Hello!

I am utilizing openapi-typescript-code-generator for a project where I generate both the client and server (python through connexion/aiohttp).

I like the path-oriented structure of creating the API spec, however, I have noticed other tools do often times not play nicely when using remote references. 

So far I have used another tool to generate the OpenAPI spec into a single file from the path-oriented structure. However, this file inlines all the specifications under the components/ folder. 

So my question is if there is a way for openapi-typescript-code-generator to output a merged OpenAPI monolith file where all schemas are defined under "components" and then referenced to with "#/components/SomeSchema" in the rest of the file?

Thanks for a great tool!

/Fredrik

**Comment by Himenon** [comment, marked as answer]
Very good question!

I am working on being able to output only Parased Open API Spec right now. I am currently making adjustments in [#46](https://github.com/Himenon/openapi-typescript-code-generator/pull/46) to achieve this. I'm planning to implement this fix in the near future.

To answer your question below, I'm a little unsure if the changes I'm making will accomplish this.

> So my question is if there is a way for openapi-typescript-code-generator to output a merged OpenAPI monolith file where all schemas are defined under "components" and then referenced to with "#/components/SomeSchema" in the rest of the file?

If you can provide a more specific example, I can respond more clearly, can you provide a sample code? The answer will probably be yes, but I would like to answer the question more precisely.

Thanks.

**Reply by fsvenson** [reply]
Simply put, I would like to generate a single YAML specification file from the path-oriented structure. Perhaps this example script can give you an idea of what I'm looking to do:
```python
#!/usr/bin/env python
import re
from pathlib import Path
from textwrap import indent

header = """\
openapi: '3.0.3'
info:
  title: Generated dummy API from path-oriented schema structure.
  version: '1.0'
  description: For use with other tools, e.g. openapi-generator.
"""
print(header)

written_parts = set()
for path in Path("components").rglob("*.yaml"):
    if "schemas" not in str(path):
        # Only bother with 'schemas' for now
        continue
    for i, part in enumerate(path.parts):
        parts = tuple(path.parts[0 : i + 1])
        if parts in written_parts:
            continue
        written_parts.add(parts)
        if not part.endswith(".yaml"):
            print(indent(f"{part}:", i * "  "))
        else:
            name = indent(Path(part).stem, i * "  ")
            print(f"{name}:")
            with open(path) as f:
                schema = f.read()
                schema = re.sub(r"\.yaml", "", schema)
                schema = re.sub(r'"([A-Z])', r'"#/components/schemas/\1', schema)
                print(indent(schema, (i + 1) * "  "))

# Add a dummy path to make the spec valid
dummy_paths = """\
paths:
  /dummy/path:
    get:
      summary:  Dummy path
      responses:
        '200':
          description: Success
          content:
            application/text:
              schema:
                type: string
                enum:
                  - Dummy"""

print(dummy_paths)
```

**Reply by Himenon** [reply]
I've used the [schema here](https://github.com/Himenon/openapi-typescript-code-generator/tree/main/test/api.test.domain) to get the code working and understand what you want to achieve. There is currently no way to faithfully reproduce the same behavior, and this library will not provide a way to faithfully reproduce it itself.

However, don't worry. I will be preparing an API that will be a part of achieving the same behavior. I can't provide this functionality right now, but the use cases you have suggested will be within the scope of future changes I envision.

**Reply by fsvenson** [reply]
OK, thank you for your answer. I will be keeping an eye out for your new API then!
